### ISSUE TITLE
cannot bind on "::" even though port is free

### ISSUE THREAD

**Issue by majewsky** [matrix-org/synapse#2030]
I set up Synapse **0.19.2** on my server ([here's the configuration that I use](https://github.com/majewsky/system-configuration/blob/92ade06411497848c3c5b94f29628ae57428386c/hologram-bethselamin-synapse.pkg.toml)). Upon first starting it, it failed with the following log output in the systemd journal:

```
Mär 18 21:27:38 bethselamin synapse[26713]: twisted: [] SynapseSite (TLS) starting on 8448
Mär 18 21:27:38 bethselamin synapse[26713]: twisted: [] Starting factory <synapse.http.site.SynapseSite instance at 0x7f061bea5758>
Mär 18 21:27:38 bethselamin synapse[26713]: twisted: [] Traceback (most recent call last):
Mär 18 21:27:38 bethselamin synapse[26713]: twisted: []   File "/usr/lib/python2.7/runpy.py", line 174, in _run_module_as_main
Mär 18 21:27:38 bethselamin synapse[26713]: twisted: []     "__main__", fname, loader, pkg_name)
Mär 18 21:27:38 bethselamin synapse[26713]: twisted: []   File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
Mär 18 21:27:38 bethselamin synapse[26713]: twisted: []     exec code in run_globals
Mär 18 21:27:38 bethselamin synapse[26713]: twisted: []   File "/usr/lib/python2.7/site-packages/synapse/app/homeserver.py", line 491, in <module>
Mär 18 21:27:38 bethselamin synapse[26713]: twisted: []     main()
Mär 18 21:27:38 bethselamin synapse[26713]: twisted: []   File "/usr/lib/python2.7/site-packages/synapse/app/homeserver.py", line 486, in main
Mär 18 21:27:38 bethselamin synapse[26713]: twisted: []     hs = setup(sys.argv[1:])
Mär 18 21:27:38 bethselamin synapse[26713]: twisted: []   File "/usr/lib/python2.7/site-packages/synapse/app/homeserver.py", line 336, in setup
Mär 18 21:27:38 bethselamin synapse[26713]: twisted: []     hs.start_listening()
Mär 18 21:27:38 bethselamin synapse[26713]: twisted: []   File "/usr/lib/python2.7/site-packages/synapse/app/homeserver.py", line 209, in start_listening
Mär 18 21:27:38 bethselamin synapse[26713]: twisted: []     self._listener_http(config, listener)
Mär 18 21:27:38 bethselamin synapse[26713]: twisted: []   File "/usr/lib/python2.7/site-packages/synapse/app/homeserver.py", line 188, in _listener_http
Mär 18 21:27:38 bethselamin synapse[26713]: twisted: []     interface=address
Mär 18 21:27:38 bethselamin synapse[26713]: twisted: []   File "/usr/lib/python2.7/site-packages/twisted/internet/posixbase.py", line 505, in listenSSL
Mär 18 21:27:38 bethselamin synapse[26713]: twisted: []     port = self.listenTCP(port, tlsFactory, backlog, interface)
Mär 18 21:27:38 bethselamin synapse[26713]: twisted: []   File "/usr/lib/python2.7/site-packages/twisted/internet/posixbase.py", line 478, in listenTCP
Mär 18 21:27:38 bethselamin synapse[26713]: twisted: []     p.startListening()
Mär 18 21:27:38 bethselamin synapse[26713]: twisted: []   File "/usr/lib/python2.7/site-packages/twisted/internet/tcp.py", line 983, in startListening
Mär 18 21:27:38 bethselamin synapse[26713]: twisted: []     raise CannotListenError(self.interface, self.port, le)
Mär 18 21:27:38 bethselamin synapse[26713]: twisted: [] twisted.internet.error.CannotListenError: Couldn't listen on :::8448: [Errno 98] Address already in use.
```

The odd thing is, port 8448 is free. When I do an `ncat --listen :: 8448`, it binds to that port just fine. I have removed `::` from the `bind_addresses` for now, but I suppose that will cause trouble when a federating server wants to connect over IPv6 and gets stuck in the connection timeout. Any idea what's going on there?

**Comment by APwhitehat** [comment]
Just a random idea , could you change your config file from
in the line 58
`bind_addresses: [ '0.0.0.0', '::' ]`
to
`bind_addresses: ['::', '0.0.0.0' ]`

and see if `Couldn't listen on :::8448: [Errno 98] Address already in use.`
changes to
`Couldn't listen on 0.0.0.0:8448: [Errno 98] Address already in use.`

**Comment by majewsky** [comment]
Indeed so. Which is kind of odd. I thought of the same, but ruled it out because `ncat --listen :: 8448` also worked when Synapse was already listening on `0.0.0.0:8448`.

I just checked: With `bind_addresses: ['::']`, I can connect via IPv6 and also via IPv4 loopback (using `curl -4 https://bethselamin.de` with an /etc/hosts entry of `127.0.0.1 bethselamin.de`), but not via the server's public IPv4 address. So I'm still as confused as before, maybe even more so. ¯\_(ツ)_/¯

**Comment by APwhitehat** [comment]
And I suppose you are able to connect via server public IPv4 when you have
`bind_addresses: [ '0.0.0.0']` only ?
This is really weird 😕

**Comment by majewsky** [comment]
Exactly.

Does Matrix do anything fishy with Twisted? Because, if you're just passing the bind address into Twisted, I am tempted to file a bug report with them as well.

**Comment by APwhitehat** [comment]
I'm sorry , I don't know the code base so well as to comment on this .

However , in C/C++ 
We can seperating IPv6 bindings from IPv4 by using 
`int one=1;`
`setsockopt(sockfd, IPPROTO_IPV6, IPV6_V6ONLY, &one, sizeof(one));`

I couldn't find anything of this sort in the twisted documentation (I looked only once)
Maybe they avoid this segregation which in turn makes things messy. (hence  twisted has the bug )

**Comment by majewsky** [comment]
The thing that confuses me is this: If it were so simple, wouldn't people have run into this problem much sooner? `bind_addresses: [ '0.0.0.0', '::']` is precisely what `--generate-config` produces by default.

**Comment by ArchangeGabriel** [comment]
Hum… Here it generated:
```
    # Local addresses to listen on.
    # This will listen on all IPv4 addresses by default.
    bind_addresses:
      - '0.0.0.0'
      # Uncomment to listen on all IPv6 interfaces
      # N.B: On at least Linux this will also listen on all IPv4
      # addresses, so you will need to comment out the line above.
      # - '::'
```

I should try the IPv6 one, but my current access doesn’t have IPv6 (but my server does).

**Comment by majewsky** [comment]
Ah, so I edited that. Thanks for clarifying. Anyway, I just reloaded my firewall rules to be extra certain that that's not the problem. And `curl -4 https://bethselamin.de:8448` (going via the public v4 IP from DNS) didn't work at first, but while trying to isolate the problem, it started working. I have no idea what is going on there, but I will count this as a misconfiguration somewhere on my side. Thanks for the advice, everyone.

**Comment by Myridium** [comment]
I'm having the same issue. And same behaviour as the OP. If I erase `'::'` then I can connect on IPv4. But adding `'::'` gives an error that the port is already bound, with error 98. It appears to be exactly the same issue as the OP.

```
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,150 - twisted - 192 - INFO -  - SynapseSite (TLS) starting on 8448
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,153 - twisted - 192 - ERROR -  - Traceback (most recent call last):
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,153 - twisted - 192 - ERROR -  -   File "/usr/lib/python3.8/site-packages/twisted/internet/tcp.py", line 1361, in startListening
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,154 - twisted - 192 - ERROR -  -     skt.bind(addr)
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,154 - twisted - 192 - ERROR -  - OSError: [Errno 98] Address already in use
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,154 - twisted - 192 - ERROR -  -
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,154 - twisted - 192 - ERROR -  - During handling of the above exception, another exception occurred:
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,154 - twisted - 192 - ERROR -  -
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,155 - twisted - 192 - ERROR -  - Traceback (most recent call last):
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,155 - twisted - 192 - ERROR -  -   File "/usr/lib/python3.8/site-packages/synapse/app/_base.py", line 278, in start
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,155 - twisted - 192 - ERROR -  -     hs.start_listening(listeners)
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,155 - twisted - 192 - ERROR -  -   File "/usr/lib/python3.8/site-packages/synapse/app/homeserver.py", line 284, in start_listening
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,155 - twisted - 192 - ERROR -  -     self._listening_services.extend(self._listener_http(config, listener))
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,155 - twisted - 192 - ERROR -  -   File "/usr/lib/python3.8/site-packages/synapse/app/homeserver.py", line 134, in _listener_http
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,156 - twisted - 192 - ERROR -  -     ports = listen_ssl(
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,156 - twisted - 192 - ERROR -  -   File "/usr/lib/python3.8/site-packages/synapse/app/_base.py", line 197, in listen_ssl
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,156 - twisted - 192 - ERROR -  -     check_bind_error(e, address, bind_addresses)
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,156 - twisted - 192 - ERROR -  -   File "/usr/lib/python3.8/site-packages/synapse/app/__init__.py", line 51, in check_bind_error
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,156 - twisted - 192 - ERROR -  -     raise e
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,156 - twisted - 192 - ERROR -  -   File "/usr/lib/python3.8/site-packages/synapse/app/_base.py", line 194, in listen_ssl
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,156 - twisted - 192 - ERROR -  -     reactor.listenSSL(port, factory, context_factory, backlog, address)
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,157 - twisted - 192 - ERROR -  -   File "/usr/lib/python3.8/site-packages/twisted/internet/posixbase.py", line 522, in listenSSL
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,157 - twisted - 192 - ERROR -  -     port = self.listenTCP(port, tlsFactory, backlog, interface)
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,157 - twisted - 192 - ERROR -  -   File "/usr/lib/python3.8/site-packages/twisted/internet/posixbase.py", line 495, in listenTCP
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,157 - twisted - 192 - ERROR -  -     p.startListening()
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,157 - twisted - 192 - ERROR -  -   File "/usr/lib/python3.8/site-packages/twisted/internet/tcp.py", line 1363, in startListening
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,157 - twisted - 192 - ERROR -  -     raise CannotListenError(self.interface, self.port, le)
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,158 - twisted - 192 - ERROR -  - twisted.internet.error.CannotListenError: Couldn't listen on :::8448: [Errno 98] Address already in use.
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,158 - synapse.handlers.deactivate_account - 196 - INFO - user_parter_loop-0 - User parter finished: stopping
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,159 - synapse.handlers.presence - 320 - INFO - presence.on_shutdown-0 - Performing _on_shutdown. Persisting 0 unpersisted changes
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,160 - synapse.handlers.presence - 333 - INFO - presence.on_shutdown-0 - Finished _on_shutdown
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,177 - twisted - 192 - CRITICAL -  - Unhandled error in Deferred:
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,177 - twisted - 192 - CRITICAL -  -
Jun 01 21:02:08 putnam synapse[498405]: Traceback (most recent call last):
Jun 01 21:02:08 putnam synapse[498405]:   File "/usr/lib/python3.8/site-packages/twisted/internet/tcp.py", line 1361, in startListening
Jun 01 21:02:08 putnam synapse[498405]:     skt.bind(addr)
Jun 01 21:02:08 putnam synapse[498405]: OSError: [Errno 98] Address already in use
Jun 01 21:02:08 putnam synapse[498405]: During handling of the above exception, another exception occurred:
Jun 01 21:02:08 putnam synapse[498405]: Traceback (most recent call last):
Jun 01 21:02:08 putnam synapse[498405]:   File "/usr/lib/python3.8/site-packages/synapse/app/_base.py", line 278, in start
Jun 01 21:02:08 putnam synapse[498405]:     hs.start_listening(listeners)
Jun 01 21:02:08 putnam synapse[498405]:   File "/usr/lib/python3.8/site-packages/synapse/app/homeserver.py", line 284, in start_listening
Jun 01 21:02:08 putnam synapse[498405]:     self._listening_services.extend(self._listener_http(config, listener))
Jun 01 21:02:08 putnam synapse[498405]:   File "/usr/lib/python3.8/site-packages/synapse/app/homeserver.py", line 134, in _listener_http
Jun 01 21:02:08 putnam synapse[498405]:     ports = listen_ssl(
Jun 01 21:02:08 putnam synapse[498405]:   File "/usr/lib/python3.8/site-packages/synapse/app/_base.py", line 197, in listen_ssl
Jun 01 21:02:08 putnam synapse[498405]:     check_bind_error(e, address, bind_addresses)
Jun 01 21:02:08 putnam synapse[498405]:   File "/usr/lib/python3.8/site-packages/synapse/app/__init__.py", line 51, in check_bind_error
Jun 01 21:02:08 putnam synapse[498405]:     raise e
Jun 01 21:02:08 putnam synapse[498405]:   File "/usr/lib/python3.8/site-packages/synapse/app/_base.py", line 194, in listen_ssl
Jun 01 21:02:08 putnam synapse[498405]:     reactor.listenSSL(port, factory, context_factory, backlog, address)
Jun 01 21:02:08 putnam synapse[498405]:   File "/usr/lib/python3.8/site-packages/twisted/internet/posixbase.py", line 522, in listenSSL
Jun 01 21:02:08 putnam synapse[498405]:     port = self.listenTCP(port, tlsFactory, backlog, interface)
Jun 01 21:02:08 putnam synapse[498405]:   File "/usr/lib/python3.8/site-packages/twisted/internet/posixbase.py", line 495, in listenTCP
Jun 01 21:02:08 putnam synapse[498405]:     p.startListening()
Jun 01 21:02:08 putnam synapse[498405]:   File "/usr/lib/python3.8/site-packages/twisted/internet/tcp.py", line 1363, in startListening
Jun 01 21:02:08 putnam synapse[498405]:     raise CannotListenError(self.interface, self.port, le)
Jun 01 21:02:08 putnam synapse[498405]: twisted.internet.error.CannotListenError: Couldn't listen on :::8448: [Errno 98] Address already in use.
Jun 01 21:02:08 putnam synapse[498405]: During handling of the above exception, another exception occurred:
Jun 01 21:02:08 putnam synapse[498405]: Traceback (most recent call last):
Jun 01 21:02:08 putnam synapse[498405]:   File "/usr/lib/python3.8/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
Jun 01 21:02:08 putnam synapse[498405]:     result = g.send(result)
Jun 01 21:02:08 putnam synapse[498405]:   File "/usr/lib/python3.8/site-packages/synapse/app/homeserver.py", line 425, in start
Jun 01 21:02:08 putnam synapse[498405]:     _base.start(hs, config.listeners)
Jun 01 21:02:08 putnam synapse[498405]:   File "/usr/lib/python3.8/site-packages/synapse/app/_base.py", line 298, in start
Jun 01 21:02:08 putnam synapse[498405]:     sys.exit(1)
Jun 01 21:02:08 putnam synapse[498405]: SystemExit: 1
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,182 - twisted - 192 - INFO -  - (TLS Port 8448 Closed)
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,183 - synapse.metrics - 464 - INFO -  - Collecting gc 1
Jun 01 21:02:08 putnam synapse[498405]: 2020-06-01 21:02:08,197 - twisted - 192 - INFO -  - Main loop terminated.
Jun 01 21:02:08 putnam systemd[1]: synapse.service: Succeeded.
```

**Comment by tom-zemingo** [comment]
Hi. I'm experiencing the same issue.
How did you resolve it?

**Comment by HammyHavoc** [comment]
Wondering the same thing.

**Comment by richvdh** [comment]
We do not monitor closed issues, in general. If people are still seeing this, please open a new issue with full details including the log output and your `listeners` config from `homeserver.yaml`.

(A likely workaround is to just listen on `::` rather than both `::` and `0.0.0.0`)

**Comment by richvdh** [comment]
note that

> `[ '0.0.0.0', '::']`

will not work; this should be fine though:

> `['::', '0.0.0.0']`
